x-pack/legacy/plugins/actions/server/builtin_action_types/es_index.ts

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { schema, TypeOf, Type } from '@kbn/config-schema';
+
+import { ActionType, ActionTypeExecutorOptions, ActionTypeExecutorResult } from '../types';
+
+function nullableType<V>(type: Type<V>) {
+  return schema.oneOf([type, schema.literal(null)], { defaultValue: () => null });
+}
+
+// config definition
+
+const unencryptedConfigProperties: string[] = ['index'];
+
+export type ActionTypeConfigType = TypeOf<typeof ConfigSchema>;
+
+const ConfigSchema = schema.object({
+  index: nullableType(schema.string()),
+});
+
+// params definition
+
+export type ActionParamsType = TypeOf<typeof ParamsSchema>;
+
+const IndexSingleBodySchema = schema.recordOf(schema.string(), schema.any());
+const IndexArrayBodySchema = schema.arrayOf(IndexSingleBodySchema);
+const IndexBodySchema = schema.oneOf([IndexSingleBodySchema, IndexArrayBodySchema]);
+
+type IndexArrayBodySchemaType = TypeOf<typeof IndexArrayBodySchema>;
+
+// see: https://www.elastic.co/guide/en/elastic-stack-overview/current/actions-index.html
+// - timeout not added here, as this seems to be a generic thing we want to do
+//   eventually: https://github.com/elastic/kibana/projects/26#card-24087404
+const ParamsSchema = schema.object({
+  index: schema.maybe(schema.string()),
+  doc_id: schema.maybe(schema.string()),
+  execution_time_field: schema.maybe(schema.string()),
+  refresh: schema.maybe(schema.boolean()),
+  body: IndexBodySchema,
```

Comment:
looks like we are also missing `refresh` and `transform` from the options that watcher index action supports.  Thinking that `body` isn't the best name for this param, maybe `document` instead?  Also thinking that we should make this always an array, to make it more obvious that you can pass multiples here.

Reply:
`documents` sounds good, as does always making it an array; that means `doc_id` would go away, as that was only applicable to the single document case.

`refresh` is already there.  `transform` isn't, because I figured `transform` was a general watcher thing, perhaps not appropriate for Kibana.  Guessing that if we want it here, we want it as a general capability for all actions.  And to some extent, this may intersect with the thought of using templates to generate action parameters.  And maybe what we want in the end is to use "expressions" to handle all this jazz.  Expressions being the kind of "painless" of Kibana (in terms of a general transformer thing, and it's ubiquity) ...

Reply:
@bmcconaghy so ... whatcha thinking on `transform`?  Even though I think it's possible it could be a general thing for actions, perhaps it's really most appropriate for the `index` action, since that one deal in fairly arbitrary JSON-able objects, vs the other actions which have less arbitrary params and are more "notification"-y vs "update the db"-y.

Perhaps some research is required to see how people are using it today with watcher, come up with a use case for using it today?

Reply:
Yeah don't really know how it is used today.  I suppose that the alert that was going to trigger this action could do its own transformation, so maybe it's not necessary.  Just was looking to see if we were at watcher parity and noticed this.